### ISSUE TITLE
macOS: Generating fltk.framework directory which is not a Framework

### ISSUE THREAD

**Issue by MatthiasWM** [fltk/fltk#961]
**Describe the bug**
Current CMake on macOS generates `/usr/local/FLTK.framework` which does not contain any headers or libraries, but is still picked up and preferred over `/usr/local/lib/libfltk.dylib`. Since the framework is empty, builds fail (i.e. TigerVNC) with _fltk library  not found_. I know that this a known bug and @Albrecht-S and @ManoloFLTK work on it, but I could not find an Issue attached to this. 

I am happy to help.

**To Reproduce**
`cmake -B build`
`cmake --build build`
`sudo cmake --build build --target install`
`ls -R /usr/local/FLTK.framework`

**Expected behavior**
IMHO it makes no sense to create this directory at all. If anything, just as we can configure FLTK to generate a static library, or a dynamic library, we should have an option to explicitly enabling frameworks.

If we do that, we must generate all frameworks for fltk, including fltk-images, fltk-jpeg, etc. . They should be installed in `/Library/Frameworks` if meant for everyone on the machine, or in `~/Library/Frameworks` if aimed at the current user, but not in `/usr/local`. They can also be part of the .app Package: [Apple Docs](https://developer.apple.com/library/archive/documentation/MacOSX/Conceptual/BPFrameworks/Tasks/InstallingFrameworks.html)

Frameworks vs. Posix style libraries is somewhat of an either-or decision, and CMake should not be given an opportunity to accidentally mix and match them.

**FLTK Version**
FLTK master (1.4 alpha)

**Operating System / Platform:**
macOS, any version

**Comment by Albrecht-S** [comment]
> Current CMake on macOS generates /usr/local/FLTK.framework which does not contain any headers or libraries, but is still picked up and preferred over /usr/local/lib/libfltk.dylib.

I believe the current code to generate this pseudo framework can safely be removed or replaced. This code predates my attempts to move to Modern CMake.

> we must generate all frameworks for fltk, including fltk-images, fltk-jpeg, etc. . They should be installed in /Library/Frameworks if meant for everyone on the machine, or in ~/Library/Frameworks if aimed at the current user, but not in /usr/local.

Please keep in mind that we **_must respect_** the installation path given by the user (`CMAKE_INSTALL_PREFIX`). `/usr/local` is obviously not a good choice on macOS (CMake default?) but if the user doesn't change it then so be it. There are some special framework related paths in CMake for installing relative to the install prefix but I don't remember the details. We should definitely use common conventions. TL;DR: note that I don't know the details of what Apple recommends either.

**Comment by ManoloFLTK** [comment]
> I know that this a known bug and @Albrecht-S and @ManoloFLTK work on it, but I could not find an Issue attached to this.

I believe we concluded with Albrecht this is an issue to tackle after 1.4 has been released. The present situation where FLTK does not create frameworks (apart from the badly named non-framework `/usr/local/FLTK.framework/ ` mentioned above) but creates .dylib's is OK and allows client apps to link to installed shared FLTK libraries. 

Notice that `/usr/local/FLTK.framework/ ` is necessary at present to build a client FLTK app with "modern" cmake against an installed version of FLTK shared libs. So, this badly named directory should be kept until the next step for FLTK 1.5.

After the release, we could work to make the FLTK build procedure create proper frameworks. I like very much the macOS approach where all frameworks are put inside the client bundle because this allows the app to be self sufficient and distributed without having to fiddle in the file system. This has the drawback that any FLTK client app would duplicate each FLTK framework. That's what can be seen, for example, in `/Applications/Microsoft\ Word.app/Contents/Frameworks`and  `/Applications/Microsoft\ PowerPoint.app/Contents/Frameworks`.

> I am happy to help.

Great.

**Comment by ManoloFLTK** [comment]
> Current CMake on macOS generates /usr/local/FLTK.framework which does not contain any headers or libraries, but is still picked up and preferred over /usr/local/lib/libfltk.dylib. Since the framework is empty, builds fail (i.e. TigerVNC) with fltk library not found. 

I've just tried here to build on macOS a client FLTK app with the modern CMake procedure described in `README.CMake.txt` in 2 ways
- against an installed set of static FLTK libs;
- against an installed set of dynamic FLTK libs.

In both cases, the build is successful and the app runs. The target installation directory was not specified which amounts to `/usr/local`. In both cases `/usr/local/FLTK.framework`is necessary.  The client program's `CMakeLists.txt `needs a small change between the 2 builds. The static build uses
`target_link_libraries(client-name PRIVATE fltk::images)`
The dynamic build uses
`target_link_libraries(client-name PRIVATE fltk::images-shared)`

What is the exact problem hit when building TigerVNC?

**Comment by ManoloFLTK** [comment]
@MatthiasWM @Albrecht-S It turns out the confusing name "FLTK.framework" can be avoided as follows:
- replace line 99 of `CMake/setup.cmake`
`  set(FLTK_CONFIG_PATH FLTK.framework/Resources/CMake)`
by, for example,
`  set(FLTK_CONFIG_PATH FLTK.cmake-config)`

- Use that name in the client's `CMakeLists.txt` when defining `FLTK_DIR` which indicates where the FLTK "package" can be found:
`set(FLTK_DIR "/usr/local/FLTK.cmake-config"  CACHE FILEPATH "FLTK installation or build directory")`

The result is that CMake builds correctly the client app using the installed set of FLTK libs and no non-framework directory with the `.framework` extension gets created. A small problem is that `make uninstall` doesn't delete these directories.

**Comment by MatthiasWM** [comment]
Here is what I get with the current GitHub versions of FLTK and TigerVNC. Output shortened to what I relevant. 

``` bash
% git clone https://github.com/fltk/fltk.git fltk-1.4.git
% cd fltk-1.4.git
% cmake -B build/Makefiles -DFLTK_BUILD_TEST=off -DFLTK_BUILD_SHARED_LIBS=on
% cmake --build build/Makefiles
% sudo cmake --build build/Makefiles --target uninstall
% sudo cmake --build build/Makefiles --target install
% cd ..
% git clone https://github.com/TigerVNC/tigervnc.git tigervnc.git
Fix tigervnc .vncviewer/DesktopWindow.cxx:157: error: reference to non-static member function...
% cd tigervnc.git
% cmake -B build
-- Found FLTK: /usr/local/lib/libfltk_images.dylib;/usr/local/fltk.framework
                                              bad  ~~~~~~~~~~~~~~~~~~~~~~~~~
% cmake --build build
[ 83%] Linking CXX executable vncviewer
ld: framework 'fltk' not found
clang: error: linker command failed with exit code 1 (use -v to see invocation)
% sudo rm -R /usr/local/FLTK.framework
% rm build/CMakeCache.txt
% cmake -B build
-- Found FLTK: /usr/local/lib/libfltk_images.dylib;/usr/local/lib/libfltk.dylib
                                             good  ~~~~~~~~~~~~~~~~~~~~~~~~~
% cmake --build build
% ./build/vncviewer/vncviewer
```

CMake picks up on `/usr/local/fltk.framework`, but that directory is not a framework in the macOS Framework sense of things. Under macOS, a directory with a name ending in `.framework` has a very specific internal structure, and would have to contain an `info.plist`, the FLTK header files and all libraries to make any sense.

Make stopped searching for libfltk.dylib after it already found fltk.framework, but since fltk.framework contains no library, the build fails. If it does not fail for you, you may still have leftovers from old fltk installations that may be picked up ten before `/usr/local/fltk.framework`.

**Comment by MatthiasWM** [comment]
> @MatthiasWM @Albrecht-S It turns out the confusing name "FLTK.framework" can be avoided as follows:
> 
> * replace line 99 of `CMake/setup.cmake`
>   `  set(FLTK_CONFIG_PATH FLTK.framework/Resources/CMake)`
>   by, for example,
>   `  set(FLTK_CONFIG_PATH FLTK.cmake-config)`

Thanks, I'll look into that. So there is no code to generate a Framework in the macOS sense? Good. Then we will keep it that way for now. I'll check what the best cmake compatible location for the config files would be.

Thanks for looking into this.

**Comment by ManoloFLTK** [comment]
> you may still have leftovers from old fltk installations 

I don't, because I never install FLTK, except for today's experiments. I would believe the culprit is in TigerVNC's `CMakeLists.txt` and how it describes where to find FLTK. Anyway, my other post shows how to prevent having any pseudo framework directory around.

**Comment by ManoloFLTK** [comment]
> So there is no code to generate a Framework in the macOS sense? 

Not yet. As written above, our idea was to create such code after 1.4 so it's ready for 1.5, and it's great you volunteer to do that.

**Comment by Albrecht-S** [comment]
@MatthiasWM @ManoloFLTK Thanks to both of you for taking care of this.

I agree with Manolo that we can and should(!) remove the pseudo framework from our CMake build and install the config file in a proper place instead. That way everything that needs FLTK can build properly using the `FLTKConfig.cmake` file in CMake **_CONFIG_** mode. Maybe that's what `TigerVNC` does "wrong", i.e. not using `CONFIG` mode explicitly but just using find_package(FLTK ...) ?

> Then we will keep it that way for now. I'll check what the best cmake compatible location for the config files would be.

It should probably be somewhere below the install prefix in 'share' or 'share/FLTK' or something like that. The CMake docs describe where "packages" (and config files) should be installed, i.e. where the standard search tries to find them.

**Comment by ManoloFLTK** [comment]
> Would /usr/local/lib/cmake/FLTK-1.4/ be the correct location? 

I would prefer a location somewhere in `/usr/local/share/` because the CMake-configuration file under discussion is not a library.

**Comment by Albrecht-S** [comment]
> Would /usr/local/lib/cmake/FLTK-1.4/ be the correct location? 

Maybe not, I don't think that there is **one** correct answer to this question. It comes likely closer to a "correct" answer if you replace `/usr/local` with `${CMAKE_INSTALL_PREFIX}` but this is not meant to be used in the code as-is (there are other CMake variables like `FRAMEWORK_LOCATION` (off the top of my head, not necessarily correct).

> The [CMake Docs](https://cmake.org/cmake/help/latest/command/find_package.html) are really unclear to me, but at least SDL2 and Vulkan write directories using this scheme. OpenCV writes to /usr/local/share/OpenCV/ . 

I **believe** it is common to use \<prefix>/lib for system libraries but in other cases it might be more appropriate to use \<prefix>/share as OpenCV does.

"Versioning" like `FLTK-1.4` may also be an option, IIRC CMake searches for such "versioned" directory names in some cases as well.

> Sorry, I sometimes just don't get CMake's thinking.

I don't know if anybody can follow everything `CMake` is doing. CMake has been devoloped over a long time, and adding new features has led to really confusing options for searching things etc..

If you want to look at more CMake docs, [cmake-packages](https://cmake.org/cmake/help/latest/manual/cmake-packages.7.html) would be a good starter, but probably not that alone.

For the near future (1.4.x) I would definitely prefer not to build FLTK as a framework. We should remove that non-framework stuff but keep the library-only build as it is now. Building and using macOS frameworks is really complex, as can be read in Craig Scott's book "Professional CMake". Its chapter _"25. Apple Features"_ is 33 pages long and much of the framework related stuff has been improved in very recent CMake versions and/or the author recommends to use Xcode for building. Requiring Xcode for building FLTK would something I'd like to avoid as long as possible.

In the future (1.5.x) we **_might_** add framework support and maybe only if built with Xcode - and let "normal libraries" be built for other tools like Make and Ninja, or something like that. But this needs thorough investigation.

**Comment by Albrecht-S** [comment]
@ManoloFLTK wrote:
> I would prefer a location somewhere in /usr/local/share/ because the CMake-configuration file under discussion is not a library.

After reading Manolo's reply I noticed that Matt's question about the "correct location" was ambiguous. I think we need to consider two completely different aspects:

1. The current location of the config file inside a pseudo (non-)framework must be fixed. This config file should probably be located under `<prefix>/share`.
2. Building a real framework and its location should be considered later (1.5.0).

My previous reply was more meant regarding building of a framework than the location of the config file.

**Comment by MatthiasWM** [comment]
> @MatthiasWM @Albrecht-S It turns out the confusing name "FLTK.framework" can be avoided as follows:
> 
> * replace line 99 of `CMake/setup.cmake`
>   `  set(FLTK_CONFIG_PATH FLTK.framework/Resources/CMake)`
>   by, for example,
>   `  set(FLTK_CONFIG_PATH FLTK.cmake-config)`
> * Use that name in the client's `CMakeLists.txt` when defining `FLTK_DIR` which indicates where the FLTK "package" can be found:
>   `set(FLTK_DIR "/usr/local/FLTK.cmake-config"  CACHE FILEPATH "FLTK installation or build directory")`
> 
> The result is that CMake builds correctly the client app using the installed set of FLTK libs and no non-framework directory with the `.framework` extension gets created. A small problem is that `make uninstall` doesn't delete these directories.

Sorry to be a bother. So can I execte the first part and change `FLTK_CONFIG_PATH`?

For the second part, do I need to copy that information to `README.CMake.txt` ?

**Comment by ManoloFLTK** [comment]
@MatthiasWM Here is my current opinion, but consider @Albrecht-S knows much better than I do all CMake implications any change can have:
> So can I execute the first part and change FLTK_CONFIG_PATH?

I would suggest yes using this formula:
   `set(FLTK_CONFIG_PATH share/FLTK.cmake-config)`
so that the output CMake config directory lands in $CMAKE_INSTALL_PREFIX/share/ .

> For the second part, do I need to copy that information to README.CMake.txt ?

I believe yes, without forgetting these changes apply for macOS only and installed FLTK only.

**Comment by Albrecht-S** [comment]
I'll try to look into this tomorrow and let you know what I suggest, or maybe I'll commit a change.

**Comment by rudironsonijr** [comment]
same issue here trying to build TigerVNC on CLion:

![CleanShot 2024-08-21 at 14 41 30@2x](https://github.com/user-attachments/assets/aef99a55-1db8-4195-8bcd-68dd9185f680)

**Comment by Albrecht-S** [comment]
@rudironsonijr Thanks for your comment. Does TigerVNC now officially support FLTK 1.4 (Git)?

Note that I am currently working on a fix for FLTK 1.4 which **_may_** also be backported to 1.3.10 (not sure yet). This solution will NOT generate a framework in 1.3/1.4 but will avoid installing a broken pseudo-framework (as it is now).

I need some more tests and I'm trying to find out (in the CMake docs) what the proper (recommended) storage location would be.

Creating a real macOS Framework will not be done in 1.4.0 (likely not before 1.5.0).

**Comment by ManoloFLTK** [comment]
@MatthiasWM @Albrecht-S :  I believe to understand now the origin of the pseudo framework directory, and what should FLTK do for 1.4, until real frameworks are created in 1.5 (presumably).

With the present code, FLTK installation under macOS produces directory `/usr/local/FLTK.framework/Resources/CMake/` and puts there 5 configuration files (`FLTKConfig.cmake`and friends) that are found under Linux in `/usr/local/share/fltk/`. As it is, this doesn't make sense because `/usr/local/FLTK.framework/` is not a real framework. But this makes totally sense if this is considered part of a complete procedure that does create a real framework named `FLTK.framework` containing a binary (a shared library), header files and CMake configuration files. Frameworks are useful exactly because they can bundle many things besides shared executables: header files, resources, other frameworks,…, and CMake configuration files.

But FLTK 1.4 will not use frameworks and will stay with good old shared libraries, e.g.,  `libfltk.1.4.0.dylib`. The adequate thing to do at this stage, I believe now, is to have macOS do the same as Linux does when installing FLTK, that is, set line 99 of file `CMake/setup.cmake` to
`set(FLTK_CONFIG_PATH share/fltk)`and obtain, after installation, shared FLTK libs in `/usr/local/lib/` and CMake configuration files in `/usr/local/share/fltk/`.

No need to give special macOS instructions in `README.CMake.txt`.

**Comment by MatthiasWM** [comment]
> same issue here trying to build TigerVNC on CLion:

Just delete the "framework" file and TigerVNC should compile again.

**Comment by Albrecht-S** [comment]
@ManoloFLTK wrote:
> The adequate thing to do at this stage, I believe now, is to have macOS do the same as Linux does when installing FLTK, that is, set line 99 of file CMake/setup.cmake to  `set(FLTK_CONFIG_PATH share/fltk)`

... or better yet to `set(FLTK_CONFIG_PATH ${FLTK_DATADIR}/fltk)` which is **_literally the same_** as Linux does which is eventually transformed to `share/fltk` (if using the defaults).

That's what I did previously but I still need to figure out what the proper (recommended) install (sub)directory would be for macOS because the [CMake docs mention different search paths per OS](https://cmake.org/cmake/help/latest/command/find_package.html#search-procedure). Sorry for the long delay.

... checking it ...

The search path `<prefix>/(lib/<arch>|lib*|share)/<name>*/` (for Unix in general) matches the given path `share/fltk`, hence we should be fine with this approach. I'm going to commit and push it, and then we can IMHO close this issue. Later, in 1.5, when we add real Framework support, a new issue may be opened for this.

**Comment by Albrecht-S** [comment]
FTR: Fix pushed in commit dd819a118cfaa889bec6563b6a8e1e969b91f7ee.

Note: I left the formal distinction between `macOS` and `Linux/Unix` intentionally in the code although the two assignments are now identical - for later updates, when we may need to change it for real macOS Frameworks.

Can we close this issue now?

**Comment by ManoloFLTK** [comment]
Closing this since Matthias expressed approval.
